zstd-safe/src/lib.rs

```diff
@@ -1048,15 +1084,129 @@ pub struct InBuffer<'a> {
     pub pos: usize,
 }
 
+/// Describe a resizeable bytes container like `Vec<u8>`.
+///
+/// Can start from uninitialized memory, and will be partially filled.
+///
+/// Should be implemented by a contiguous chunk of memory.
+///
+/// The main implementors are:
+/// * `Vec<u8>` and similar structures. These can start empty with a non-zero capacity, and they
+///   will be resized to cover the data written.
+/// * `[u8]` and `[u8; N]`. These must start already-initialized, and will not be resized. It will
+///   be up to the caller to only use the part that was written.
+pub trait WriteBuf {
+    /// Returns the valid data part of this container. Should only cover initialized data.
+    fn as_slice(&self) -> &[u8];
+
+    /// Returns the full capacity of this container. May include uninitialized data.
+    fn capacity(&self) -> usize;
+
+    /// Returns a pointer to the start of the data.
+    fn as_mut_ptr(&mut self) -> *mut u8;
+
+    /// Indicates that the first `n` bytes of the container have been written.
+    unsafe fn filled_until(&mut self, n: usize);
+
+    /// Call the given closure using the pointer and capacity from `self`.
+    ///
+    /// Assumes the given function returns a parseable code, which if valid, represents how many
+    /// bytes were written to `self`.
+    unsafe fn write_from<F>(&mut self, f: F) -> SafeResult
+    where
+        F: FnOnce(*mut c_void, usize) -> usize,
```

Comment:
What do you think about having `F: FnOnce(*mut c_void, usize) -> SafeResult` here? It'll mean typing a few more `parse_code(...)` instances where its used but it would make the contract here clearer I think.

Reply:
Yeaah, I was thinking about it and was lazy to type `parse_code` everywhere, but it does make it a bit clearer.